OrbitQt/opengldetect.cpp

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "opengldetect.h"
+
+#include <QOffscreenSurface>
+#include <QOpenGLContext>
+
+namespace OrbitGl {
+
+// Detects the supported version of Desktop OpenGL by requesting the most
+// recent version of OpenGL and checking what the system could provide.
+// OpenGL ES is automatically ignored.
+std::optional<OpenGlVersion> DetectOpenGlVersion() {
+  QSurfaceFormat format{};
+  format.setRenderableType(QSurfaceFormat::OpenGL);
+  format.setVersion(4, 6);
+
+  QOffscreenSurface surface{};
+  surface.setFormat(format);
+  surface.create();
+
+  if (!surface.isValid()) {
+    return std::nullopt;
+  }
+
+  QOpenGLContext glContext{};
+  glContext.setFormat(format);
+  glContext.create();
+  glContext.makeCurrent(&surface);
+
+  if (!glContext.isValid() || glContext.isOpenGLES()) {
```

Comment:
Why is OpenGLES considered as not valid?

Reply:
OpenGL ES is always available in Qt due to ANGLE. Since we want to detect if we can run Orbit on this machine we return nothing if Qt fell back to OpenGL ES

Reply:
Ok sounds good, adding a comment could be useful here I think.

Reply:
There is a comment above the function describing the whole logic. Do you think we need more?

Reply:
I added a comment explaining what's happening.